### ISSUE TITLE
Feature Planning for openHAB 5

### ISSUE THREAD

**Issue by holgerfriedrich** [openhab/openhab-distro#1693]
### (Incomplete) Development Plan for openHAB 5

**see project:
https://github.com/orgs/openhab/projects/11/views/1**

### Background

There has been a long discussion in #1689 regarding the rollout of Java 21 and the end of life of 32-bit systems. To cut a long story short, Java 21 and the EOL for 32-bit systems should be coupled with next major release, OH 5. A possible timeline is not aligned yet (and should depend on the progress on the items listed above). The earliest possible date is summer 2025 (derived from the progress on the user visible feature "support Matter").

Let's use the list above as a starting point for a discussion. It is for sure incomplete - please comment, and I will update the list. Maybe you have features in mind that could be a selling point for our OH 5 release.

**Comment by wborn** [comment]
We could also create issues for everything instead of maintaining the list here and add them to a project: https://github.com/orgs/openhab/projects

**Comment by holgerfriedrich** [comment]
@wborn I created a project containing the items mentioned above.
https://github.com/orgs/openhab/projects/11/views/1
I misused items as heading to give it some structure. Can you please check if the order is reasonable when you view it?
Another idea would be using custom OH5xxxx tags.
WDYT?

**Comment by florian-h05** [comment]
@lsiepel See https://github.com/openhab/openhab-addons/pull/17720 for the GraalJS upgrade.

**Comment by florian-h05** [comment]
I just updated the project and added some functional and visible features.

**Comment by holgerfriedrich** [comment]
Thanks, @florian-h05!

**Comment by holgerfriedrich** [comment]
Ping @openhab/maintainers, you are welcome to join a discussion and provide your thoughts on OH5 planning!
Not sure if everybody can edit the project https://github.com/orgs/openhab/projects/11/views/1. Feel free to add your comments to this issue.

**Comment by miloit** [comment]
It would be nice to also have the possibility of an esp32s3 running as voice satellite for the openhab voice recognition

**Comment by lsiepel** [comment]
Currently working on a Z-Wave JS integration and i hope it is ready for openHAB 5, no promise though. I think it would be a great alternative for our zwave implementation.

**Comment by florian-h05** [comment]
> It would be nice to also have the possibility of an esp32s3 running as voice satellite for the openhab voice recognition

Not sure what would be required for that, but I think https://github.com/openhab/openhab-core/pull/4032 which is needed for voice assistant functionality in the UI would be important for that as well. WDYT?

> Currently working on a Z-Wave JS integration and i hope it is ready for openHAB 5, no promise though.

Do you want to create an issue for that and add to the openHAB 5 feature planning project?

**Comment by miloit** [comment]
Here is a real basic MVP, but it's running. I am not a programmer so if someone has better idea please up to you

Esp32 code
`
#include <WiFi.h>
#include <driver/i2s.h>

// Wi-Fi credentials
const char* ssid = "YOUR_SSID";
const char* password = "YOUR_PASSWORD";
const int port = 12345;

// I2S configuration
#define I2S_WS 25  // Word Select (LRCLK)
#define I2S_SD 26  // Serial Data (DOUT)
#define I2S_SCK 27 // Serial Clock (BCLK)

void setup() {
  Serial.begin(115200);
  
  // Connect to Wi-Fi
  WiFi.begin(ssid, password);
  while (WiFi.status() != WL_CONNECTED) {
    delay(1000);
    Serial.println("Connecting to WiFi...");
  }
  Serial.println("Connected to WiFi");

  // Configure I2S
  i2s_config_t i2s_config = {
    .mode = (i2s_mode_t)(I2S_MODE_MASTER | I2S_MODE_RX),
    .sample_rate = 44100,              // Sample rate
    .bits_per_sample = I2S_BITS_PER_SAMPLE_16BIT,
    .channel_format = I2S_CHANNEL_FMT_ONLY_LEFT,
    .communication_format = I2S_COMM_FORMAT_I2S,
    .intr_alloc_flags = ESP_INTR_FLAG_LEVEL1, // Interrupt level 1
    .dma_buf_count = 8,
    .dma_buf_len = 512,
    .use_apll = false
  };

  i2s_pin_config_t pin_config = {
    .bck_io_num = I2S_SCK,
    .ws_io_num = I2S_WS,
    .data_out_num = I2S_PIN_NO_CHANGE,
    .data_in_num = I2S_SD
  };

  i2s_driver_install(I2S_NUM_0, &i2s_config, 0, NULL);
  i2s_set_pin(I2S_NUM_0, &pin_config);
}

void loop() {
  WiFiClient client = WiFiServer(port).available();
  
  if (client) {
    Serial.println("Client connected");
    uint8_t buffer[512];
    size_t bytes_read;

    while (client.connected()) {
      i2s_read(I2S_NUM_0, buffer, sizeof(buffer), &bytes_read, portMAX_DELAY);
      if (bytes_read > 0) {
        client.write(buffer, bytes_read);
      }
    }

    client.stop();
    Serial.println("Client disconnected");
  }
}
`

Java Programm

`
import javax.sound.sampled.*;
import java.io.BufferedInputStream;
import java.io.InputStream;
import java.net.ServerSocket;
import java.net.Socket;

public class AudioReceiver {
    private static final int PORT = 12345;
    private static final int BUFFER_SIZE = 512;

    public static void main(String[] args) {
        try (ServerSocket serverSocket = new ServerSocket(PORT)) {
            System.out.println("Listening for connections on port " + PORT);
            Socket socket = serverSocket.accept();
            System.out.println("Client connected");

            // Audio format: 44.1kHz, 16-bit, mono, signed, big-endian
            AudioFormat format = new AudioFormat(44100, 16, 1, true, true);
            DataLine.Info info = new DataLine.Info(SourceDataLine.class, format);
            SourceDataLine audioLine = (SourceDataLine) AudioSystem.getLine(info);
            audioLine.open(format);
            audioLine.start();

            InputStream input = new BufferedInputStream(socket.getInputStream());
            byte[] buffer = new byte[BUFFER_SIZE];

            int bytesRead;
            while ((bytesRead = input.read(buffer)) != -1) {
                audioLine.write(buffer, 0, bytesRead);
            }

            audioLine.drain();
            audioLine.close();
            socket.close();
            System.out.println("Connection closed");

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
`

Basically a websocket port is necessary for the audio sink

Also an idea you can look it up on rhasspy 
https://github.com/Romkabouter/ESP32-Rhasspy-Satellite
And 
https://github.com/rhasspy/rhasspy

**Comment by miloit** [comment]
Another nice feature will be

https://github.com/openhab/openhab-webui/pull/2884

**Comment by lolodomo** [comment]
@holgerfriedrich : is this topic for everybody over the world asks what he would like in OH 5 ?!

**Comment by holgerfriedrich** [comment]
@lolodomo I tend to say yes, partly - especially if we get suggestions from regular contributors. Initially this was more intended for the maintainers. As we use the "project" linked in the post, there is no way to comment there.
The project itself should be editable by maintainers and visible to everybody.

Though, my recommendation is to focus on key features and main selling points in this issue. Topics should be mentioned here in the comments if considered relevant as key feature or selling point. Discussions and implementation details should be separated out in new issues - otherwise we will see too much traffic here and everybody removes the subscription for this planning issue.

The general wishlist discussion for future releases should be discussed in the community forum.

Just my 2 cents.

**Comment by mstormi** [comment]
Github is for OH contributors ONLY, the policy can be read just below my post.
That is, developers and maintainers ONLY, not OH users or even anybody.

@miloit please cease and move over to the forum.

**Comment by mherwege** [comment]
I think removing the default persistence service and replacing it by an easier initial configuration for it would also make sense for OH 5.0. This has been discussed before. I created a core issue for it: https://github.com/openhab/openhab-core/issues/4470

**Comment by digitaldan** [comment]
Hi, The upcoming[ Matter Binding ](https://community.openhab.org/t/matter-binding/161138) i'm working on requires a NodeJS runtime.  The binding is able to download and run this for most systems with the exception of our Alpine Linux image due to libc issues.  The binding is smart enough to try and use a system installed version (if it meets the min requirement), so it would be nice to have our Docker builds include this.

Normally i would not modify our container builds for a specific binding, but IMHO Matter is going to be a big deal this year and hopefully a big part of the 5.0 release.

**Comment by florian-h05** [comment]
I see no harm in adding Node.js to with a installed size of 45 MiB (see https://pkgs.alpinelinux.org/package/edge/main/x86/nodejs) to the Alpine Linux Docker image, IMO Matter is important enough to justify doing that :+1: 

I have just created a PR for that: https://github.com/openhab/openhab-docker/pull/454

**Comment by lsiepel** [comment]
> Hi, The upcoming[ Matter Binding ](https://community.openhab.org/t/matter-binding/161138) i'm working on requires a NodeJS runtime. The binding is able to download and run this for most systems with the exception of our Alpine Linux image due to libc issues. The binding is smart enough to try and use a system installed version (if it meets the min requirement), so it would be nice to have our Docker builds include this.
> 
> Normally i would not modify our container builds for a specific binding, but IMHO Matter is going to be a big deal this year and hopefully a big part of the 5.0 release.

Does this also affect openHABian? Or is node.js already available @mstormi ?

**Comment by ecdye** [comment]
NodeJS is already installed on the base system, or at least has been in the past for frontail. We can make sure that it is included in the base system going forward as well.

**Comment by dilyanpalauzov** [comment]
Will SQLite in OH 5 work with 32 bits?

My reading is that major reason to switch to 64 bit is that newer GraalVM only works on 64 bit.  However I prefer to use 32 bit, so my question is what will and what will not work in 32bit mode.  In particular the SQLite JDBC / org.xerial.sqlite-jdbc  bundle has a org/sqlite/native/ directory with these files:
```
./FreeBSD/aarch64/libsqlitejdbc.so:       ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, stripped
./FreeBSD/x86_64/libsqlitejdbc.so:        ELF 64-bit LSB shared object, x86-64, version 1 (FreeBSD), dynamically linked, stripped
./FreeBSD/x86/libsqlitejdbc.so:           ELF 32-bit LSB shared object, Intel 80386, version 1 (FreeBSD), dynamically linked, stripped
./Linux-Musl/aarch64/libsqlitejdbc.so:    ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, stripped
./Linux-Musl/x86_64/libsqlitejdbc.so:     ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, stripped
./Linux-Musl/x86/libsqlitejdbc.so:        ELF 32-bit LSB shared object, Intel 80386, version 1 (SYSV), dynamically linked, stripped
./Linux/aarch64/libsqlitejdbc.so:         ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, BuildID[sha1]=6a1f51a26405282633bc7b6708e2fa30eb59093a, stripped
./Linux/arm/libsqlitejdbc.so:             ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, BuildID[sha1]=ede611d5aa2147d8f5795c444b094b38c8c4fe88, stripped
./Linux/x86_64/libsqlitejdbc.so:          ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, stripped
./Linux/ppc64/libsqlitejdbc.so:           ELF 64-bit LSB shared object, 64-bit PowerPC or cisco 7500, OpenPOWER ELF V2 ABI, version 1 (SYSV), dynamically linked, BuildID[sha1]=d55d03e535dae27740d8b64c0e8369d49f4f74d3, stripped
./Linux/x86/libsqlitejdbc.so:             ELF 32-bit LSB shared object, Intel 80386, version 1 (SYSV), dynamically linked, stripped
./Linux/armv6/libsqlitejdbc.so:           ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, BuildID[sha1]=b03d5abef69089c42138cf267fb334c71dc2cbdc, stripped
./Linux/armv7/libsqlitejdbc.so:           ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, BuildID[sha1]=1d74f3bb57d3b3309d452ab6be8f15b39fa15b83, stripped
./Linux-Android/aarch64/libsqlitejdbc.so: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, stripped
./Linux-Android/arm/libsqlitejdbc.so:     ELF 32-bit LSB shared object, ARM, EABI5 version 1 (SYSV), dynamically linked, stripped
./Linux-Android/x86_64/libsqlitejdbc.so:  ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, stripped
./Linux-Android/x86/libsqlitejdbc.so:     ELF 32-bit LSB shared object, Intel 80386, version 1 (SYSV), dynamically linked, stripped
./Windows/aarch64/sqlitejdbc.dll:         PE32+ executable (DLL) (GUI) Aarch64, for MS Windows, 7 sections
./Windows/x86_64/sqlitejdbc.dll:          PE32+ executable (DLL) (console) x86-64 (stripped to external PDB), for MS Windows, 11 sections
./Windows/x86/sqlitejdbc.dll:             PE32 executable (DLL) (console) Intel 80386 (stripped to external PDB), for MS Windows, 9 sections
./Windows/armv7/sqlitejdbc.dll:           PE32 executable (DLL) (GUI) ARMv7 Thumb, for MS Windows, 6 sections
./Mac/aarch64/libsqlitejdbc.jnilib:       Mach-O 64-bit arm64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|NO_REEXPORTED_DYLIBS>
./Mac/x86_64/libsqlitejdbc.jnilib:        Mach-O 64-bit x86_64 dynamically linked shared library, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL|NO_R

```
Will the Linux 32 binaries stay in the bundle in OH 5 or will they be removed?

**Comment by mstormi** [comment]
Expect them to get removed.
And please respect that GitHub is not for openHAB users. Put your questions and discussions over to the community forum.

**Comment by florian-h05** [comment]
> My reading is that major reason to switch to 64 bit is that newer GraalVM only works on 64 bit.

The reason is libraries like GraalVM, some persistence drivers, JVM availability on 32-bit ARM (IIRC there is only one JVM available, but not one from any big JVM vendor), and support effort. We will drop support for 32-bit, if you get it work, good for you, but any bug reports will be considered as invalid and also don’t expect help in general if you are on 32-bit.

**Comment by florian-h05** [comment]
A few weeks ago I came up with another feature idea for openHAB 5:
Better configuration as code support — this would allow to provide Items and Things from scripting languages, with the benefit you could write some functions to generate a whole lot of them.

I have created https://github.com/openhab/openhab-core/pull/4513, which implements handling of the entities lifecycle based on the script lifecycle in core, scripts can just use the provided registry delegates to provide Items and Things. 

WDYT?

**Comment by florian-h05** [comment]
I partly knew about JRubys functionality here, JS Scripting also had it partly implemented some time ago. 
My intention behind the core PR is to make it as easy as possible for the automation languages so they don’t have to implement and register their own providers.

**Comment by wborn** [comment]
I think we can close this issue now that the feature freeze has happened? 🥶 

See: https://community.openhab.org/t/openhab-5-0-milestone-builds/162685/5
